src/base/utils/gzip.cpp

```diff
@@ -29,129 +29,195 @@
+    const unsigned int chunkSize = 128 * 1024;
 
     z_stream strm {};
     strm.zalloc = Z_NULL;
     strm.zfree = Z_NULL;
     strm.opaque = Z_NULL;
-    strm.next_in = reinterpret_cast<const Bytef *>(data.constData());
-    strm.avail_in = uInt(data.size());
-    strm.next_out = reinterpret_cast<Bytef *>(tmpBuf.data());
-    strm.avail_out = BUFSIZE;
+
+    char *in = new char[chunkSize];
+    char *out = new char[chunkSize];
 
     // windowBits = 15 + 16 to enable gzip
     // From the zlib manual: windowBits can also be greater than 15 for optional gzip encoding. Add 16 to windowBits
     // to write a simple gzip header and trailer around the compressed data instead of a zlib wrapper.
-    int result = deflateInit2(&strm, level, Z_DEFLATED, (15 + 16), 9, Z_DEFAULT_STRATEGY);
-    if (result != Z_OK)
-        return {};
-
-    QByteArray output;
-    output.reserve(deflateBound(&strm, data.size()));
+    int ret = deflateInit2(&strm, level, Z_DEFLATED, (15 + 16), 9, Z_DEFAULT_STRATEGY);
+    if (ret != Z_OK)
+        return false;
 
-    // feed to deflate
-    while (strm.avail_in > 0)
+    int flush;
+    do
     {
-        result = deflate(&strm, Z_NO_FLUSH);
-
-        if (result != Z_OK)
+        Q_ASSERT(source.isReadable());
+        const qsizetype readBytes = source.read(in, chunkSize);
+        if (readBytes == -1)
         {
             deflateEnd(&strm);
-            return {};
+            return false;
         }
 
-        output.append(tmpBuf.data(), (BUFSIZE - strm.avail_out));
-        strm.next_out = reinterpret_cast<Bytef *>(tmpBuf.data());
-        strm.avail_out = BUFSIZE;
-    }
-
-    // flush the rest from deflate
-    while (result != Z_STREAM_END)
-    {
-        result = deflate(&strm, Z_FINISH);
+        flush = (source.atEnd()) ? Z_FINISH : Z_NO_FLUSH;
+        strm.avail_in = readBytes;
+        strm.next_in = reinterpret_cast<const Bytef *>(in);
 
-        output.append(tmpBuf.data(), (BUFSIZE - strm.avail_out));
-        strm.next_out = reinterpret_cast<Bytef *>(tmpBuf.data());
-        strm.avail_out = BUFSIZE;
-    }
+        do
+        {
+            strm.avail_out = chunkSize;
+            strm.next_out = reinterpret_cast<Bytef *>(out);
+
+            ret = deflate(&strm, flush);
+            Q_ASSERT(ret != Z_STREAM_ERROR);
+
+            Q_ASSERT(dest.isWritable());
+            const qsizetype have = chunkSize - strm.avail_out;
+            if (dest.write(out, have) == -1)
+            {
+                deflateEnd(&strm);
+                return false;
+            }
+        } while (strm.avail_out == 0);
+        Q_ASSERT(strm.avail_in == 0);
+    } while (flush != Z_FINISH);
+    Q_ASSERT(ret == Z_STREAM_END);
 
     deflateEnd(&strm);
-
-    if (ok) *ok = true;
-    return output;
+    return true;
 }
 
-QByteArray Utils::Gzip::decompress(const QByteArray &data, bool *ok)
+QByteArray Utils::Gzip::compress(const QByteArray &data, const int level, bool *ok)
```

Comment:
I would like to ask to not update this function as I have plans to minimize the buffer copy operations. This function is in hot path for webapi responses so it would make sense to optimize it IMO.